File: app/normalize.py
"""Raw-to-enriched transaction normalization."""

from __future__ import annotations

from dataclasses import dataclass

import psycopg


@dataclass(frozen=True)
class NormalizeResult:
    created_count: int


def normalize_raw_transactions(database_url: str) -> NormalizeResult:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            result = connection.execute(
                """
                INSERT INTO enriched_transactions (
                    raw_transaction_id,
                    category_id,
                    is_income,
                    is_transfer,
                    is_savings,
                    is_fixed_cost,
                    is_variable_cost,
                    needs_review,
                    classification_method,
                    classification_confidence,
                    classification_reason
                )
                SELECT
                    raw_transactions.id,
                    unknown_category.id,
                    false,
                    false,
                    false,
                    false,
                    true,
                    true,
                    'normalized',
                    0,
                    'Created from raw transaction; deterministic classification not yet applied.'
                FROM raw_transactions
                CROSS JOIN categories AS unknown_category
                LEFT JOIN enriched_transactions
                    ON enriched_transactions.raw_transaction_id = raw_transactions.id
                WHERE unknown_category.name = 'Unknown'
                    AND enriched_transactions.id IS NULL
                RETURNING enriched_transactions.id
                """
            ).fetchall()

    return NormalizeResult(created_count=len(result))

File: src/server/transactions/normalize.ts
import type pg from "pg";

import { sql, toQuery } from "../db/sql.ts";

export async function normalizePendingTransactions(client: pg.PoolClient): Promise<number> {
  const result = await client.query<{ id: string }>(
    toQuery(sql`
      INSERT INTO enriched_transactions (
        raw_transaction_id, category_id, needs_review, classification_method,
        classification_confidence, classification_reason
      )
      SELECT
        raw_transactions.id,
        categories.id,
        true,
        'uncategorized',
        0,
        'Created during normalization; classification has not run yet.'
      FROM raw_transactions
      CROSS JOIN categories
      WHERE categories.name = 'Unknown'
        AND NOT EXISTS (
          SELECT 1
          FROM enriched_transactions
          WHERE enriched_transactions.raw_transaction_id = raw_transactions.id
        )
      RETURNING id
    `),
  );

  return result.rowCount ?? 0;
}
